### PR TITLE
Backfill hasher speed boost: using a tree to create a grouping for package files hash

### PR DESCRIPTION
**Pull request by kenotron**
# Description

TLDR: this change drastically improves the hash time. My Microsoft Teams hashing went from >60s to 2s.

Currently, we have a bottleneck in calculating package hashes. The slow down is from matching the correct repo file hashes with the files that are inside a package. The problem is that we do these slow things:

1. for every package, we take all the files in a repo and try to filter for package and THEN redo this for every single package (# of files * # of package operations)
2. the path.join() operation is very slow, and we use it to do the comparison

# Details

## `repoInfo.ts`
To address the problems, we need a way to group all the `repoHashes` by package paths. We first construct an intermediate lookup tree for all the package folders in this shape:

```ts
interface PathNode = { [key: string]: PathNode };
```

Using that, we then lookup the package folders for all the files present in the repo, and then collating all the file hashes into:

```ts
type FileHashTuple = [string, string];
type PackageHashes = Record<string, FileHashTuple[]>;

// PackageHashes has package relative paths from root as the key, and a [file path, file hash] tuple
```
## `hashOfFiles.ts`, `hashOfFiles.test.ts`
The `generateHashOfFiles` function now has a "slow" and a "fast" path. The "slow" path is retained to allow the same behavior as the old implementation when the monorepo has a malformed shape (when the function is called inside a folder that really doesn't belong to a workspace). The newer implementation requires the package folder to actually be a package inside a workspace (known from a `getWorkspaces` call).

The monorepo test actually used an incomplete and malformed fixture to test, resulting in call to the slow path. A new test is added that uses the proper monorepo fixture to make sure the hashing is correct.

## Chore

I've had to bump eslint & ts+eslint plugins to accommodate the newer TS 4.x series. Otherwise, it was complaining about undefined "map" or some such error. The 8.x eslint also did away with separated prettier plugins - so they're all folded into one.

### CODE REVIEW

File: packages/hasher/src/hashOfFiles.ts [hunk 1]
```diff
@@ -13,29 +12,39 @@ import { RepoInfo } from "./repoInfo";
...
+    // Slow old path: if files are not clearly inside a package (mostly the case for malformed monorepos, like tests)
+    const normalized = path.normalize(packageRoot) + sep;
 
-  files.sort((a, b) => a.localeCompare(b));
+    const files: string[] = Object.keys(repoHashes).filter((f) =>
+      path.join(root, f).includes(normalized)
+    );
 
-  const hashes: string[] = [];
+    files.sort((a, b) => a.localeCompare(b));
```
Comment (dzearing) [inline packages/hasher/src/hashOfFiles.ts:40]: is this sort necessary?

Comment (dzearing) [inline packages/hasher/src/hashOfFiles.ts:40]: Add a comment for the purpose of this sort. I think it's so that your hashes are in a predictable order.

File: packages/hasher/src/repoInfo.ts [hunk 2]
```diff
@@ -29,24 +31,97 @@ function searchRepoInfoCache(packageRoot: string) {
   }
 }
 
+function createPackageHashes(
```
Comment (dzearing) [inline packages/hasher/src/repoInfo.ts:34]: Optional: consider moving this to it's own file and adding a test. Not required, but nice to have small testable units.

**Review by dzearing — APPROVED** [review]
